### PR TITLE
RFC: Breaking API Change: drivers: can: rework support for manual bus-off recovery

### LINKED ISSUE #69459: RFC: Breaking API Change: drivers: can: rework support for manual bus-off recovery
## Introduction

Many CAN controllers support two different modes for recovering from bus-off state:
- Automatic recovery after 128 occurrences of 11 consecutive recessive bits have been monitored on the bus.
- Manual recovery initiation, after which the CAN controller still waits until 128 occurrences of 11 consecutive recessive bits have been monitored on the bus.

The Zephyr CAN controller API supports both modes through the the `CONFIG_CAN_AUTO_BUS_OFF_RECOVERY` Kconfig option, which is enabled by default. The manual recovery API function `can_recover()` is only functional when `CONFIG_CAN_AUTO_BUS_OFF_RECOVERY` is disabled.

### Problem description

Even though ISO11898-1 §6.15 states that no CAN node should automatically recover from bus-off events, manual recovery initiation is not consistently supported by the CAN controllers/drivers available in Zephyr.

This leads to incoherent behavior between drivers with `CONFIG_CAN_AUTO_BUS_OFF_RECOVERY=n`. Drivers supporting manual recovery initiation disable automatic bus-off recovery at driver initialization, while the ones not supporting manual recovery initiation keep it enabled regardless of `CONFIG_CAN_AUTO_BUS_OFF_RECOVERY=n` (there is no other option) and return `-ENOTSUP` from the mandatory `can_recover()` API function.

An overview of the currently supported CAN controllers along with their support manual/automatic bus-off recovery is shown in the table below.

| Driver compatible | Automatic Recovery | Manual Recovery |
| - | - | - |
| atmel,sam0-can | X | X |
| atmel,sam-can | X | X |
| espressif,esp32-twai | X | X |
| infineon,xmc4xxx-can | X | |
| kvaser,pcican | X | X |
| microchip,mcp2515 | X | |
| microchip,mcp251xfd | X | |
| nuvoton,numaker-canfd | X | X |
| nxp,flexcan | X | X |
| nxp,lpc-mcan | X | X |
| nxp,s32-canxl | X | X |
| renesas,rcar-can | X | X |
| st,stm32-bxcan | X | X |
| st,stm32-fdcan | X | X |
| st,stm32h7-fdcan | X | X |
| ti,tcan4x5x | X | X |
| zephyr,can-loopback | (X) | |
| zephyr,fake-can | (X) | |
| zephyr,native-linux-can | (X) | |

### Proposed change

Since all CAN controllers seem to support automatic recovery (for any future drivers for hardware without this hardware capability this can easily be implemented in the driver), change the Zephyr CAN controller API policy to:
- Always enable automatic bus recovery upon driver initialization, regardless of Kconfig options. Since CAN controllers are initialized in "stopped" state, no unwanted bus-off recovery will be started at this point.
- Invert and rename the Kconfig `CONFIG_CAN_AUTO_BUS_OFF_RECOVERY`, which is enabled by default, to `CONFIG_CAN_MANUAL_RECOVERY_MODE`, which is disabled by default. Enabling `CONFIG_CAN_MANUAL_RECOVERY_MODE=y` enables support for the `can_recover()` API function and a new manual recovery mode (see next bullet). Keeping this guarded by Kconfig allows keeping the flash footprint down for applications not using manual bus-off recovery.
- Introduce a new CAN controller operational mode `CAN_MODE_MANUAL_RECOVERY`. Support for this is only enabled if `CONFIG_CAN_MANUAL_RECOVERY_MODE=y`. Having this as a mode allows applications to inquire whether the CAN controller supports manual recovery mode via the `can_get_capabilities()` API function and either fail or rely on automatic recovery - and it allows CAN controller drivers not supporting manual recovery mode to fail early in `can_set_mode()` during application startup instead of failing when `can_recover()` is called at a later point in time.

## Detailed RFC

Please see PR #69460.

### Proposed change (Detailed)

Please see PR #69460.

### Dependencies

None

### Concerns and Unresolved Questions

None so far.

## Alternatives

None so far.

### PR DESCRIPTION
**Pull request by henrikbrixandersen**
Since all CAN controllers drivers seem to support automatic recovery (for any future drivers for hardware without this hardware capability this can easily be implemented in the driver), change the Zephyr CAN controller API policy to:

- Always enable automatic bus recovery upon driver initialization, regardless of Kconfig options. Since CAN controllers are initialized in "stopped" state, no unwanted bus-off recovery will be started at this point.

- Invert and rename the Kconfig CONFIG_CAN_AUTO_BUS_OFF_RECOVERY, which is enabled by default, to CONFIG_CAN_MANUAL_RECOVERY_MODE, which is disabled by default. Enabling CONFIG_CAN_MANUAL_RECOVERY_MODE=y enables support for the can_recover() API function and a new manual recovery mode (see next bullet). Keeping this guarded by Kconfig allows keeping the flash footprint down for applications not using manual bus-off recovery.

- Introduce a new CAN controller operational mode CAN_MODE_MANUAL_RECOVERY. Support for this is only enabled if CONFIG_CAN_MANUAL_RECOVERY_MODE=y. Having this as a mode allows applications to inquire whether the CAN controller supports manual recovery mode via the can_get_capabilities() API function and either fail or rely on automatic recovery - and it allows CAN controller drivers not supporting manual recovery mode to fail early in can_set_mode() during application startup instead of failing when can_recover() is called at a later point in time.

RFC: #69459

### CODE REVIEW

**Comment by henrikbrixandersen** [comment]
RFC ready for review.

**Review by martinjaeger — DISMISSED** [review]
Looks good.

Only got one minor non-blocking comment which may be applied to some other drivers as well.

**Review by dleach02 — DISMISSED** [review]

**Review by manuargue — APPROVED** [review]

**Comment by congnguyenhuu** [comment]
Look good to me.
But I see that nxp s32 canxl driver is missing capability for manual recovery mode. Please update this and check this in other drivers.@henrikbrixandersen 
https://github.com/zephyrproject-rtos/zephyr/blob/8992d05b0fde9c8437a0191a6d4e973634ebf918/drivers/can/can_nxp_s32_canxl.c#L153

**Review by manuargue — CHANGES_REQUESTED** [review]
missing capability for manual recovery mode in NXP S32 CAXL driver as pointed out above

File: drivers/can/can_mcan.c [hunk 1]
```diff
@@ -350,22 +354,24 @@ int can_mcan_stop(const struct device *dev)
...
-		return -ENOTSUP;
+	if (IS_ENABLED(CONFIG_CAN_MANUAL_RECOVERY_MODE)) {
+		supported |= CAN_MODE_MANUAL_RECOVERY;
 	}
-#else  /* CONFIG_CAN_FD_MODE */
-	if ((mode & ~(CAN_MODE_LOOPBACK | CAN_MODE_LISTENONLY)) != 0U) {
+
+	if (IS_ENABLED(CONFIG_CAN_FD_MODE)) {
+		supported |= CAN_MODE_FD;
+	}
```
[the remark this answers is not in the document]
Comment (henrikbrixandersen) [reply drivers/can/can_mcan.c:369]: Thanks, yes - I too noticed that pattern when converting the drivers. I didn't want to mix that change into this PR, but I'm preparing to move this generic check into the `can_set_mode()` implementation function itself.

**Comment by henrikbrixandersen** [comment]
> Look good to me. But I see that nxp s32 canxl driver is missing capability for manual recovery mode. Please update this and check this in other drivers.@henrikbrixandersen

Nicely spotted, thanks. I have updated the patch.

**Comment by henrikbrixandersen** [comment]
Rebased on `main` to resolve merge conflicts in `doc/releases/migration-guide-3.7.rst` and `drivers/can/can_stm32_bxcan.c`.

**Review by martinjaeger — APPROVED** [review]

**Review by manuargue — APPROVED** [review]

**Pull request merged**